### PR TITLE
chore: add query-frontend option to select request headers in query logs

### LINKED ISSUE #3342: Add query frontend stats logs to tempo
**Is your feature request related to a problem? Please describe.**

I would like to see the queries being executed against Loki, and additional information in them so I can use Loki to track query performance, track down users whose queries failed, etc. 

**Describe the solution you'd like**

Implementing similar log lines as the `query stats` already present in Loki and Mimir query frontend.
* https://github.com/grafana/loki/blob/main/pkg/lokifrontend/frontend/transport/handler.go#L185
* https://github.com/grafana/mimir/blob/main/pkg/frontend/transport/handler.go#L241

**Describe alternatives you've considered**

n/a

**Additional context**

This was discussed in the Tempo Community Call. It was suggested something like that is present already, Since I couldn't find any log lines like this I opened this issue. I would be also willing to work on implementing this.

### PR DESCRIPTION
**Pull request by jmichalek132**
**What this PR does**:

Adding feature present in mimir and Loki, specifically https://github.com/grafana/mimir/pull/5030 and https://github.com/grafana/loki/issues/11422.

Adds a config option to the query-frontend to specify a list of request headers to include in query logs.

For example, setting -query-frontend.log-query-request-headers="X-Grafana-Org-Id" and sending a query with X-Grafana-Org-Id:1 results in query log lines that include header_x_grafana_org_id=1.

**Which issue(s) this PR fixes**:
Fixes #3342

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by jmichalek132** [comment]
For the pipeline failure, the test fails looks unrelated to my change:
```
tenant=0 component=remote level=info remote_name=ce2f78 url=http://127.0.0.1:35569/receive msg="Scraped metadata watcher stopped"
    instance_test.go:140: 
        	Error Trace:	/home/runner/work/tempo/tempo/modules/generator/storage/instance_test.go:140
        	            				/home/runner/work/tempo/tempo/modules/generator/storage/instance_test.go:340
        	            				/opt/hostedtoolcache/go/1.21.6/x64/src/runtime/asm_amd64.s:1650
        	Error:      	Received unexpected error:
        	            	write /tmp/TestInstance_multiTenancy3811487438/001/0/wal/00000000: file already closed
        	Test:       	TestInstance_multiTenancy
tenant=0 component=remote level=info remote_name=ce2f78 url=http://127.0.0.1:35569/receive msg="Remote storage stopped."
```

For the linting I ran `make lint` locally as instructed by the contributing guide, but it came up with many complaints for code I didn't touch, so unsure how to address that.

File: modules/frontend/handler.go [hunk 1]
```diff
@@ -81,35 +86,40 @@ func (f *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		f.post(ctx, resp, orgID, elapsed, err)
 	}
 
+	logMessage := append([]interface{}{
+		"tenant", orgID,
+		"method", r.Method,
+		"traceID", traceID,
+		"url", r.URL.RequestURI(),
+		"duration", elapsed.String(),
+	})
```
Comment (mapno) [inline modules/frontend/handler.go:95]: This should get rid of the lint error

```suggestion
	logMessage := []interface{}{
		"tenant", orgID,
		"method", r.Method,
		"traceID", traceID,
		"url", r.URL.RequestURI(),
		"duration", elapsed.String(),
	}
```

File: modules/frontend/handler.go [hunk 2]
```diff
@@ -38,19 +41,21 @@ type (
...
 type handler struct {
-	roundTripper http.RoundTripper
-	logger       log.Logger
-	post         handlerPostHook
-	pre          handlerPreHook
+	roundTripper           http.RoundTripper
+	logger                 log.Logger
+	post                   handlerPostHook
+	pre                    handlerPreHook
+	LogQueryRequestHeaders flagext.StringSliceCSV
```
Comment (mapno) [inline modules/frontend/handler.go:48]: This can be lowercase, no need to export it

```suggestion
	logQueryRequestHeaders flagext.StringSliceCSV
```

**Review by mapno — CHANGES_REQUESTED** [review]
LGTM, only to minor things.

**Comment by joe-elliott** [comment]
> For the pipeline failure, the test fails looks unrelated to my change:

Don't worry about this. CI flakiness. We can ignore this error.

**Review by mapno — APPROVED** [review]
LGTM. Thanks!

**Comment by mapno** [comment]
Ups, it seems the `logQueryRequestHeaders` wasn't correctly updated everywhere. When that's fixed I'll merge cc/ @jmichalek132

**Comment by jmichalek132** [comment]
> Ups, it seems the `logQueryRequestHeaders` wasn't correctly updated everywhere. When that's fixed I'll merge cc/ @jmichalek132

Sorry for the delay, fixed now.

**Pull request merged**